adcc/estimate_memory.py

```diff
@@ -0,0 +1,170 @@
+##
+## ---------------------------------------------------------------------
+import numpy as np
+
+from .MoSpaces import MoSpaces
+from .AdcMethod import AdcMethod
+
+
+def estimate_n_floats(method, n_o, n_c, n_v, max_subspace=0,
+                      restricted=False):
+    if not isinstance(method, AdcMethod):
+        method = AdcMethod(method)
+
+    eriimport = "vv"
+    count = {
+        "oo": 0, "ov": 0, "vv": 0,
+        "oooo": 0, "ooov": 0, "ovov": 0,
+        "oovv": 0, "ovvv": 0, "vvvv": 0,
+    }
+    sizes = {
+        "o": n_o,
+        "c": n_c,
+        "v": n_v,
+        "oo": n_o * n_o / 2,
+        "ov": n_o * n_o,
+        "vv": n_v * n_v / 2,
+        "oc": n_o * n_c,
+        "cc": n_c * n_c / 2,
+        "cv": n_c * n_v,
+        "oooo": (n_o**2 + n_o + 1) * (n_o**2 + n_o) / 8,
+        "ooov": (n_o**2 + n_o) * n_o * n_v / 2,
+        "oovv": (n_o**2 + n_o) * (n_v**2 + n_v) / 4,
+        "ovov": (n_o * n_v + 1) * (n_o * n_v) / 2,
+        "ovvv": n_o * n_v * (n_v**2 + n_v) / 2,
+        "vvvv": (n_v**2 + n_v + 1) * (n_v**2 + n_v) / 8,
+        "cccc": (n_o**2 + n_o + 1) * (n_o**2 + n_o) / 8,
+        "oocv": (n_o**2 + n_o) * n_c * n_v / 2,
+        "ocov": n_o * n_c * n_o * n_v,
+        "occv": n_o * n_c * n_c * n_v,
+        "ocvv": n_o * n_c * (n_v**2 + n_v) / 2,
+        "cccv": (n_c**2 + n_c) * n_c * n_v / 2,
+        "ccvv": (n_c**2 + n_c) * (n_v**2 + n_v) / 4,
+        "cvcv": (n_c * n_v + 1) * (n_c * n_v) / 2,
+        "cvvv": n_c * n_v * (n_v**2 + n_v) / 2,
+    }
+
+    if method.is_core_valence_separated:
+        raise NotImplementedError("CVS not yet implemented")
+    else:
+        count["oo"] += 4  # Fock + coefficients + TDMs
+        count["ov"] += 6  # Fock + coefficients + TDMs + df
+        count["vv"] += 4  # Fock + coefficients + TDMs
+
+    # Computed states (singles block)
+    count["ov"] += max_subspace * 3
+
+    if method.level >= 1:
+        count["oo"] += 1  # MP2 diffdm
+        count["ov"] += 1  # MP2 diffdm
+        count["vv"] += 1  # MP2 diffdm
+
+        count["ooov"] += 1  # ERI
+        count["oovv"] += 2  # ERI + T2
+        count["ovov"] += 1  # ERI
+        count["ovvv"] += 1  # ERI
+        eriimport = "ovvv"
+
+    if method.level >= 2:
+        count["oooo"] += 1  # ERI
+        count["vvvv"] += 1  # ERI
+        count["oovv"] += 4  # t2eri + TD2
+
+        # Even though vvvv is needed, the import
+        # relevant during iteration is ovvv for ADC(2)
+        eriimport = "ovvv"
+
+        # Computed states (doubles block)
+        count["ovov"] += max_subspace * 3
+
+    if method.name in ["adc2", "adc2x"]:
+        count["oo"] += 1  # I1 intermediate
+        count["vv"] += 1  # I2 intermediate
+        eriimport = "vvvv"
+
+    if method.level >= 3:
+        count["ovov"] += 1  # M11 intermediate
+        count["ovvv"] += 3  # t2eri + Pia intermediate
+        count["ooov"] += 3  # t2eri + Pib intermediate
+        eriimport = "vvvv"
```

Comment:
Probably we should also count the oooo and vvvv block if properties are requested?

Reply:
Good point, but this PR is far away from such micro-optimisation at the moment, unfortunately ... the results are still incredibly unreliable.